### ISSUE TITLE
FSDP2 does not work with custom Tensor wrapper objects

### ISSUE THREAD

**Issue by riku-sakamoto** [pytorch/pytorch#168143]
### 🐛 Describe the bug

Hi! First of all, thank you for your great work maintaining PyTorch.

## Issue Overview

When using `Tensor` wrapper type object as introduced in [the manual](https://docs.pytorch.org/docs/stable/notes/extending.html#extending-torch-with-a-tensor-wrapper-type), weight parameters are not updated through FSDP2 training.
Here is the example. The loss value does not change through training when using `MetadataTensor` as input tensor object.

```python
import os
from datetime import timedelta

import torch
import torch.distributed as dist
import torch.multiprocessing as mp
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.fsdp import fully_shard


class MetadataTensor(object):
    # https://docs.pytorch.org/docs/stable/notes/extending.html#extending-torch-with-a-tensor-wrapper-type
    def __init__(self, data, metadata=None, **kwargs):
        self._t = torch.as_tensor(data, **kwargs)
        self._metadata = metadata

    def __repr__(self):
        return "Metadata:\n{}\n\ndata:\n{}".format(self._metadata, self._t)

    @classmethod
    def __torch_function__(cls, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}
        metadatas = tuple(a._metadata for a in args if hasattr(a, "_metadata"))
        args = [getattr(a, "_t", a) for a in args]
        assert len(metadatas) > 0
        ret = func(*args, **kwargs)
        return MetadataTensor(ret, metadata=metadatas[0])

    def sum(self) -> torch.Tensor:
        return self._t.sum()


def main(rank: int, world_size: int, tcp_port_num: int):
    # --- 1. setup distributed environment
    os.environ["RANK"] = str(rank)
    os.environ["WORLD_SIZE"] = str(world_size)
    os.environ["LOCAL_RANK"] = str(rank)
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = str(tcp_port_num)
    device = torch.device(f"cuda:{rank}")
    dist.init_process_group(
        backend="nccl",
        rank=rank,
        world_size=world_size,
        timeout=timedelta(seconds=30),
    )
    mesh = init_device_mesh("cuda", (world_size,))

    # --- 2. define model, optimizer, data
    model = torch.nn.Linear(10, 1).to(device)
    fsdp_kwargs = {
        "mesh": mesh,
        "reshard_after_forward": False,
    }
    model = fully_shard(model, **fsdp_kwargs)
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-5)

    # NOTE: If we use MetadataTensor, the loss value does not change through training.
    # If we use normal torch.Tensor, the loss value changes as expected. 
    # fixed_x = torch.rand((3, 10), device=device)
    fixed_x = MetadataTensor(
        torch.rand((3, 10), device=device), metadata="input tensor"
    )

    # --- 3. training loop
    for _ in range(3):
        model.train()
        loss = model(fixed_x).sum()
        loss.backward()

        optimizer.step()
        optimizer.zero_grad()
        if rank == 0:
            print(f"{loss.item()=}")

    torch.distributed.destroy_process_group()


if __name__ == "__main__":

    world_size = torch.cuda.device_count()
    tcp_port_num = 32867
    mp.spawn(
        main,
        args=(world_size, tcp_port_num),
        nprocs=torch.cuda.device_count(),
        join=True,
    )

```

Output

```
loss.item()=0.5710310935974121
loss.item()=0.5710310935974121
loss.item()=0.5710310935974121
```

## Rationale

It seems that the following code causes this behavior. It allows only `torch.Tensor` objects as input tensor objects.
Thus, `RegisterPostBackwardFunction.apply` is not called for `Tensor` wrapper type object, and the gradients are not properly handled in FSDP2.
If it is not intentional design, `torch.overrides.is_tensor_like` check might be more appropriate here. If it makes sense, I am happy to open PR.

https://github.com/pytorch/pytorch/blob/6fc430644b1357fab03a03619576fef7197ac60e/torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py#L721

### Versions

```
PyTorch version: 2.9.1+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: AlmaLinux 8.10 (Cerulean Leopard) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-28)
Clang version: Could not collect
CMake version: version 3.19.6
Libc version: glibc-2.28

Python version: 3.10.13 (main, Apr 30 2024, 14:25:40) [GCC 8.5.0 20210514 (Red Hat 8.5.0-18)] (64-bit runtime)
Python platform: Linux-4.18.0-553.77.1.el8_10.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 11.1.105
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe

Nvidia driver version: 580.105.08
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:         x86_64
CPU op-mode(s):       32-bit, 64-bit
Byte Order:           Little Endian
CPU(s):               96
On-line CPU(s) list:  0-47
Off-line CPU(s) list: 48-95
Thread(s) per core:   1
Core(s) per socket:   24
Socket(s):            2
NUMA node(s):         4
Vendor ID:            GenuineIntel
CPU family:           6
Model:                106
Model name:           Intel(R) Xeon(R) Gold 6342 CPU @ 2.80GHz
Stepping:             6
CPU MHz:              2800.000
CPU max MHz:          3500.0000
CPU min MHz:          800.0000
BogoMIPS:             5600.00
Virtualization:       VT-x
L1d cache:            48K
L1i cache:            32K
L2 cache:             1280K
L3 cache:             36864K
NUMA node0 CPU(s):    0-11
NUMA node1 CPU(s):    12-23
NUMA node2 CPU(s):    24-35
NUMA node3 CPU(s):    36-47
Flags:                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall n
x pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx
 smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowpre
fetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc
_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512
vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_wind
ow hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_cl
ear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] mypy==1.18.2
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] torch==2.9.1
[pip3] torchvision==0.24.1
[pip3] triton==3.5.1
[conda] Could not collect

```
